.github/workflows/gradle.yml

```diff
@@ -1,33 +1,54 @@
 name: gradle-ci
 
-on: [ push, pull_request, workflow_dispatch ]
+on:
+  push:
+    branches:
+      - master
+  pull_request:
+    branches:
+      - master
+  workflow_dispatch:
+
+concurrency:
+  group: "${{ github.workflow }} @ ${{ github.event.pull_request.head.label || github.head_ref || github.ref }}"
+  cancel-in-progress: true
 
 jobs:
   build:
-    runs-on: ubuntu-latest
-
-    env:
-      GRADLE_OPTS: "-Dorg.gradle.jvmargs=-XX:MaxMetaspaceSize=512m"
-
     strategy:
       matrix:
         java-version: [ 11, 17, 20 ]
+        os:
+          - macos-latest
+          - ubuntu-latest
+          - windows-latest
+      fail-fast: false
+    runs-on: ${{ matrix.os }}
+    env:
+      GRADLE_OPTS: "-Dorg.gradle.jvmargs=-XX:MaxMetaspaceSize=512m"
 
     steps:
       - uses: actions/checkout@v3
-        with:
-          fetch-depth: 0
```

Comment:
`fetch-depth: 0` will fetch all history for all branches and tags, which I assume is not necessary, but maybe I'm wrong?